runtime/near-evm-runner/src/types.rs

```diff
@@ -0,0 +1,95 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use near_vm_errors::VMLogicError;
+use near_vm_logic::types::AccountId;
+use std::io::{Read, Write};
+
+pub type RawAddress = [u8; 20];
+pub type RawHash = [u8; 32];
+pub type RawU256 = [u8; 32];
+
+pub type Result<T> = std::result::Result<T, VMLogicError>;
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct AddressArg {
+    pub address: RawAddress,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct ViewCallArgs {
+    pub sender: RawAddress,
+    pub address: RawAddress,
+    pub amount: RawU256,
+    pub args: Vec<u8>,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct GetStorageAtArgs {
+    pub address: RawAddress,
+    pub key: RawHash,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct WithdrawArgs {
+    pub account_id: AccountId,
+    pub amount: RawU256,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct TransferArgs {
+    pub address: RawAddress,
+    pub amount: RawU256,
+}
+
+impl BorshSerialize for ViewCallArgs {
+    fn serialize<W: Write>(&self, writer: &mut W) -> std::io::Result<()> {
```

Comment:
why this cannot be derived?

Reply:
Because last argument doesn't have a length. Borsh requires every array to have a length encoded.

Reply:
Got it. What if reorganize it as non-last argument?

Reply:
and what then?...

Reply:
Oh i think i misread, you mean problem is not it is the `last` that needs an end but is borsh cannot derive serialize vector. But that's not true, borsh can encode vector:
```
use borsh::{BorshDeserialize, BorshSerialize};

#[derive(Debug, Eq, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct ViewCallArgs {
    pub args: Vec<u8>,
}

fn main() {
    let a = ViewCallArgs { args: vec![0,1,2] };
    let encoded_a = a.try_to_vec().unwrap();
    let decoded_a = ViewCallArgs::try_from_slice(&encoded_a).unwrap();
    assert_eq!(a, decoded_a);
}
```

borsh spec:
```
Dynamic sized array | vec_type: "Vec<" ident '>' | repr(len() as u32)                 for el in x                    repr(el as ident)

```